fullstack-cert/python-projects/abstraction-workshop/main.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC, abstractmethod
+from typing import List
+
+class Product:
+    def __init__(self, name: str, price: float):
+        self.name = name
+        self.price = price
+
+class DiscountStrategy(ABC):
+    @abstractmethod
+    def is_applicable(self, product: Product, user_tier: str) -> bool:
+        pass
+    @abstractmethod
+    def apply_discount(self, product: Product) -> float:
+        pass
+
+class PercentageDiscount(DiscountStrategy):
+    def __init__(self, percent: float):
+        self.percent = percent
+
+    def is_applicable(self, product: Product, user_tier: str) -> bool:
+        return True
+
+    def apply_discount(self, product: Product) -> float:
+        return product.price * (1 - self.percent / 100)
+
+class FixedAmountDiscount(DiscountStrategy):
+    def __init__(self, amount: float):
+        self.amount = amount
+
+    def is_applicable(self, product: Product, user_tier: str) -> bool:
+        return product.price > self.amount
```

Comment:
Here make so that any final price is fine, even 0.01
Maybe something like...

```suggestion
        return product.price*0.9 > self.amount
```

in this case the final price can't be lower than 10% of the original price